feature: Add async HTTP request support

scroix opened the pull request:
Resolves issue #38 from 2015(!) by adding async HTTP request support. Recipes which use URL requests can now rely on callbacks for async responses without blocking the main thread

- Uses `CompletableFuture` via Apache HTTP Client 5.x (the sync path now runs on the same async core)
- Adds an `async=True` flag to nodetoolkit, preserving full backward compatibility with existing synchronous calls
- Provides callback-based response handling via `complete` and `error`; without an `error` callback, failures raise into the node's console rather than disappearing
- Keeps the HttpClient 4 behaviours recipes depend on: transparent gzip/deflate decompression, in-memory responses under the 150 MB cap, and a `content` of None for bodyless responses (HEAD/204/304)

<img width="1698" height="882" alt="2025-07-19_eeJMWJQZ@2x" src="https://github.com/user-attachments/assets/0864835c-3ca8-4058-a50f-00cc13bc2f66" />

scroix committed:
Add Apache HTTP components 5.x dependencies

scroix committed:
Implement Apache5NodelHttpClient with async

  - Add async client instance within Apache5NodelHttpClient
  - Implement makeRequestAsync and makeSimpleRequestAsync methods
  - Handle API differences between HTTP Components 4.x and 5.x
  - Add proper resource management with graceful shutdown for async client
  - Add error handling and connection management
  - Update authentication and SSL context handling for compatibility

scroix committed:
Refactor Apache5NodelHttpClient

  - Cut code size by ~25%
  - Implement lazy init, only creating sync or async clients when needed
  - Consolidate error handling for both sync and async
  - Split functionality into small, focused methods
  - Attempt to maintain Java 8 compatibility by using constructs

scroix committed:
chore(http-client): simplify connection manager and SSL setup, enforce per-request timeouts

- Replace reflection-based max-connection
- Extract SSLContext initialisation
- Honor per-request connect/read timeouts
- Clean up unused imports and suppress the “unused” warning
- Bump project source/target compatibility to Java 11.

scroix committed:
Added asynchronous HTTP request function `get_url_async` to `nodetoolkit`

scroix committed:
Refactored `nodetoolkit` to ensure thread-safe callback handling for `CompletableFuture` responses

scroix committed:
Added unit tests for `getURLAsync` method to validate signature and callback handling

scroix committed:
Unified `get_url` and `get_url_async` into a single method supporting both synchronous and asynchronous modes.

scroix committed:
Removed unused `AsyncHttpClientTestStub` file.

scroix committed:
Removed Java source and target compatibility settings from build configuration.

scroix commented:
I've noticed I didn't give enough consideration to the changes to `ApacheNodelHttpClient` in order to achieve this result and will spend some more effort ensuring the transition from httpclient4 to httpclient5 is smooth.

scroix committed:
refactor: Consolidate Apache HTTP clients into a single asynchronous implementation.

scroix committed:
minor: Update class docstring to remove specific Apache HTTP client version reference

scroix committed:
style: Remove trailing whitespaces in `nodetoolkit.py` for cleanup

scroix committed:
Remove redundant comments and unused imports from `ApacheNodelHttpClient`

scroix committed:
desc: Cleaner method clarity

scroix committed:
refactor: Extract SSL config setup back into `prepareForNoSSL` method

scroix committed:
minor: Reorganise `ApacheNodelHttpClient` for further clarity

scroix committed:
Merge branch 'museumsvictoria:dev' into feature/async_getURL

scroix committed:
refactor: Introduce `SmartResponseConsumer` for memory vs. file streaming

scroix committed:
cleanup: Remove unused imports, redundant code

scroix committed:
minor: update comments, and clarify interface obligations

scroix committed:
minor: Update Apache HTTP client dependencies to latest versions

scroix commented:
I think this is finally ready for a review 😫

I've [tested](https://gist.github.com/scroix/0c055cb2f44e316455d799b6d4f68b27) a variety of use cases of `get_url` to ensure nothing is broken, but I can't of caught them all.

  - ✅ Sync requests
  - ✅ Async requests
  - ✅ POST/PUT/PATCH/DELETE all working correctly
  - ✅ Status codes and content properly populated
  - ✅ SSL verification
  - ✅ Proper streaming consumer
  - ✅ `java.io.IOException: Too big - HTTP response over 157286400 bytes is not allowed` on large files

I _didn't_ test a proxy, that's all.

justparking commented:
Hey @scroix - belated update... nice work! I wanted to let you know I haven't been able to fully review and test drive it yet but very keen to start soon, possibly today.

scroix committed:
Merge branch 'dev' into feature/async_getURL

scroix committed:
Ignore 1xx responses in async HTTP client
- Updated ApacheNodelHttpClient.informationResponse() to ignore, not throw, enabling normal 1xx (e.g. 100-Continue) handling.

scroix committed:
Merge branch 'feature/async_getURL' of github.com:scroix/nodel into feature/async_getURL

scroix committed:
Fix async HTTP client dropping Content-Type on body-less requests

scroix committed:
Preserve async HTTP client exceptions instead of wrapping in IOException

scroix committed:
Give async HTTP executor longer to drain before shutdownNow()

scroix committed:
Wait for active async HTTP requests to finish before force shutdown

scroix committed:
Prevent new async HTTP requests once client is shutting down

scroix committed:
Guard async HTTP submissions against null client shutdowns

scroix commented:
Just revisiting this as I'm keen to have the PR merged. I've addressed a couple of small issues

- wait for in-flight async requests to complete before forcing shutdown
- can't submit new requests once client is shutting down
- fix Content-Type headers being dropped without a body
- fix 1xx request regression (they should be ignored)

justparking commented:
Hey @scroix - thanks for the updates. I had been meaning update this PR with some of my findings. I set up the tests a while ago but never got around to share with you fully because I was setting up more and more tests.

Anyway, here's my feedback:

My concern is that when you use the async call it has a signifcant impact on the state of the underlying thread collection.

This is just one concern, shown below where there are a bunch of branch new dispatch thread created and there is no effort to use a much smaller threadpool, at least initially.

On a heavily loaded nodel server it would severly tax the server (memory and likely CPU) to have to create so many threads for each and every node that use's the async calls, which in the long run, is what we would ideally like to have every single recipe use.

So I'm hoping you can utilise nodel's thread-pooling framework in a way some of the other modules do.

Although I haven't tested for it here, we need to also double-check what happens threading-wise when the nodes are shutdown.

Anyway, take a look.

**Current release: One node using `get_url("...google.com")`:**
```
=== Thread Groups (Sorted by Name) @ Mon Nov 03 18:27:20 AEDT 2025 ===
 Total Active Threads: 50

 [ 2] 10_78_0_15_probe_resp_receiver
 [ 2] 172_29_208_1_probe_resp_receiver
 [ 1] Attach Listener
 [ 1] Common-Cleaner
 [ 1] Finalizer
 [ 1] NanoHttpd Main Listener
 [ 1] NanoHttpd Request Processor (#12)
 [ 1] NanoHttpd Request Processor (#14)
 [ 1] NanoHttpd Request Processor (#15)
 [ 1] NanoHttpd Request Processor (#16)
 [ 1] NanoHttpd Request Processor (#17)
 [ 1] Reference Handler
 [ 1] Signal Dispatcher
 [ 1] Thread-Dump-Sorted-Logger
 [ 1] main
 [ 1] nodel.diagnostics
 [ 1] nodel_timer_0
 [ 1] nodel_timer_1
 [ 1] nodel_timer_10
 [ 1] nodel_timer_11
 [ 1] nodel_timer_12
 [ 1] nodel_timer_13
 [ 1] nodel_timer_14
 [ 1] nodel_timer_15
 [ 1] nodel_timer_16
 [ 1] nodel_timer_17
 [ 1] nodel_timer_18
 [ 1] nodel_timer_19
 [ 1] nodel_timer_2
 [ 1] nodel_timer_3
 [ 1] nodel_timer_4
 [ 1] nodel_timer_5
 [ 1] nodel_timer_6
 [ 1] nodel_timer_7
 [ 1] nodel_timer_8
 [ 1] nodel_timer_9
 [ 1] org.nodel.core.ChannelServerSocket_000
 [ 1] org.python.google.common.base.internal.Finalizer
 [ 1] pool_Discovery_0
 [ 1] pool_Discovery_1
 [ 1] pool_Discovery_2
 [ 1] pool_Dynamic nodes_0
 [ 1] pool_Dynamic nodes_1
 [ 1] pool_Nodel WebSocket interceptor_0
 [ 1] pool_Nodel WebSocket interceptor_1
 [ 1] pool_Nodel WebSocket interceptor_2
 [ 1] pool_Nodel host_0
 [ 1] pool_Nodel host_1

=== End Grouped Thread Dump ===
```

**Async release: One node using `get_url("...google.com", async=True)`:**
```
=== Thread Groups (Sorted by Name) @ Mon Nov 03 18:34:03 AEDT 2025 ===
 Total Active Threads: 75

 [ 2] 10_78_0_15_probe_resp_receiver
 [ 2] 172_29_208_1_probe_resp_receiver
 [ 1] Attach Listener
 [ 1] Common-Cleaner
 [ 1] Finalizer
 [ 1] NanoHttpd Main Listener
 [ 1] NanoHttpd Request Processor (#10)
 [ 1] NanoHttpd Request Processor (#11)
 [ 1] NanoHttpd Request Processor (#13)
 [ 1] NanoHttpd Request Processor (#14)
 [ 1] NanoHttpd Request Processor (#2)
 [ 1] NanoHttpd Request Processor (#3)
 [ 1] Reference Handler
 [ 1] Signal Dispatcher
 [ 1] Thread-Dump-Sorted-Logger
 [ 1] httpclient-dispatch-1
 [ 1] httpclient-dispatch-10
 [ 1] httpclient-dispatch-11
 [ 1] httpclient-dispatch-12
 [ 1] httpclient-dispatch-13
 [ 1] httpclient-dispatch-14
 [ 1] httpclient-dispatch-15
 [ 1] httpclient-dispatch-16
 [ 1] httpclient-dispatch-17
 [ 1] httpclient-dispatch-18
 [ 1] httpclient-dispatch-19
 [ 1] httpclient-dispatch-2
 [ 1] httpclient-dispatch-20
 [ 1] httpclient-dispatch-3
 [ 1] httpclient-dispatch-4
 [ 1] httpclient-dispatch-5
 [ 1] httpclient-dispatch-6
 [ 1] httpclient-dispatch-7
 [ 1] httpclient-dispatch-8
 [ 1] httpclient-dispatch-9
 [ 1] httpclient-main-1
 [ 1] main
 [ 1] nodel.diagnostics
 [ 1] nodel_timer_0
 [ 1] nodel_timer_1
 [ 1] nodel_timer_10
 [ 1] nodel_timer_11
 [ 1] nodel_timer_12
 [ 1] nodel_timer_13
 [ 1] nodel_timer_14
 [ 1] nodel_timer_15
 [ 1] nodel_timer_16
 [ 1] nodel_timer_17
 [ 1] nodel_timer_18
 [ 1] nodel_timer_19
 [ 1] nodel_timer_2
 [ 1] nodel_timer_3
 [ 1] nodel_timer_4
 [ 1] nodel_timer_5
 [ 1] nodel_timer_6
 [ 1] nodel_timer_7
 [ 1] nodel_timer_8
 [ 1] nodel_timer_9
 [ 1] org.nodel.core.ChannelServerSocket_000
 [ 1] org.python.google.common.base.internal.Finalizer
 [ 1] pool-1-thread-1
 [ 1] pool_Discovery_0
 [ 1] pool_Discovery_1
 [ 1] pool_Discovery_2
 [ 1] pool_Dynamic nodes_0
 [ 1] pool_Dynamic nodes_1
 [ 1] pool_Nodel WebSocket interceptor_0
 [ 1] pool_Nodel WebSocket interceptor_1
 [ 1] pool_Nodel WebSocket interceptor_2
 [ 1] pool_Nodel host_0
 [ 1] pool_Nodel host_1
 [ 1] pool_Toolkit_0
 [ 1] pool_Toolkit_1

=== End Grouped Thread Dump ===
```

**Async release: Two nodes using `get_url("...google.com", async=True)`:**
```
=== Thread Groups (Sorted by Name) @ Mon Nov 03 18:35:44 AEDT 2025 ===
 Total Active Threads: 102

 [ 2] 10_78_0_15_probe_resp_receiver
 [ 2] 172_29_208_1_probe_resp_receiver
 [ 1] Attach Listener
 [ 1] Common-Cleaner
 [ 1] Finalizer
 [ 1] NanoHttpd Main Listener
 [ 1] NanoHttpd Request Processor (#10)
 [ 1] NanoHttpd Request Processor (#11)
 [ 1] NanoHttpd Request Processor (#13)
 [ 1] NanoHttpd Request Processor (#15)
 [ 1] NanoHttpd Request Processor (#17)
 [ 1] NanoHttpd Request Processor (#19)
 [ 1] NanoHttpd Request Processor (#21)
 [ 1] NanoHttpd Request Processor (#22)
 [ 1] NanoHttpd Request Processor (#23)
 [ 1] NanoHttpd Request Processor (#24)
 [ 1] NanoHttpd Request Processor (#26)
 [ 1] NanoHttpd Request Processor (#3)
 [ 1] Reference Handler
 [ 1] Signal Dispatcher
 [ 1] Thread-Dump-Sorted-Logger
 [ 2] httpclient-dispatch-1
 [ 2] httpclient-dispatch-10
 [ 2] httpclient-dispatch-11
 [ 2] httpclient-dispatch-12
 [ 2] httpclient-dispatch-13
 [ 2] httpclient-dispatch-14
 [ 2] httpclient-dispatch-15
 [ 2] httpclient-dispatch-16
 [ 2] httpclient-dispatch-17
 [ 2] httpclient-dispatch-18
 [ 2] httpclient-dispatch-19
 [ 2] httpclient-dispatch-2
 [ 2] httpclient-dispatch-20
 [ 2] httpclient-dispatch-3
 [ 2] httpclient-dispatch-4
 [ 2] httpclient-dispatch-5
 [ 2] httpclient-dispatch-6
 [ 2] httpclient-dispatch-7
 [ 2] httpclient-dispatch-8
 [ 2] httpclient-dispatch-9
 [ 2] httpclient-main-1
 [ 1] main
 [ 1] nodel.diagnostics
 [ 1] nodel_timer_0
 [ 1] nodel_timer_1
 [ 1] nodel_timer_10
 [ 1] nodel_timer_11
 [ 1] nodel_timer_12
 [ 1] nodel_timer_13
 [ 1] nodel_timer_14
 [ 1] nodel_timer_15
 [ 1] nodel_timer_16
 [ 1] nodel_timer_17
 [ 1] nodel_timer_18
 [ 1] nodel_timer_19
 [ 1] nodel_timer_2
 [ 1] nodel_timer_3
 [ 1] nodel_timer_4
 [ 1] nodel_timer_5
 [ 1] nodel_timer_6
 [ 1] nodel_timer_7
 [ 1] nodel_timer_8
 [ 1] nodel_timer_9
 [ 1] org.nodel.core.ChannelServerSocket_000
 [ 1] org.python.google.common.base.internal.Finalizer
 [ 1] pool-2-thread-1
 [ 1] pool_Discovery_0
 [ 1] pool_Discovery_1
 [ 1] pool_Discovery_2
 [ 1] pool_Dynamic nodes_0
 [ 1] pool_Dynamic nodes_1
 [ 1] pool_Nodel WebSocket interceptor_0
 [ 1] pool_Nodel WebSocket interceptor_1
 [ 1] pool_Nodel WebSocket interceptor_2
 [ 1] pool_Nodel host_0
 [ 1] pool_Nodel host_1
 [ 1] pool_Toolkit_0
 [ 1] pool_Toolkit_1

=== End Grouped Thread Dump ===
```

scroix committed:
refactor: Constrain async HTTP thread usage with shared pool
- Limit IO reactor threads to 2
- Replace per-instance cached ExecutorService with shared ThreadPool
- Share single 16-thread pool across all nodes for async submissions
- Remove executor shutdown logic

justparking commented:
At the very least, I reckon we could swap out...

```java
  _executor = Executors.newCachedThreadPool();
```

...with use of a large static thread-pool just for HTTP, something like:
```java
  private static ThreadPool s_threadPool = new ThreadPool("HTTP client", 256); 
```

But maybe there are others that need to be set too, one or more of:
```java
  // CUSTOM EXECUTORS
  .setIOReactorExecutor(ioReactorExecutor)        // 1. Custom I/O Reactor Thread Pool (for NIO event loop)
  .setConnectionManagerExecutor(connMgrExecutor)  // 2. Custom Connection Manager Executor (lease/release, eviction)
  .setUserExecutor(userCallbackExecutor)          // 3. Custom User Callback Executor (your response handlers)
```

FYI: I'm in the middle of some work to implement "fenced" threadpools where the nodes operate within their own individual contraints and not these package-level contraints. It is designed to prevent nodes from unintentionally interferring with others nodes yet still get the performance and small-footprint benefits of using a global thread-pool. I'm also tidying up how the naming of the thread-pools so it clear what does what.

scroix committed:
minor: Rename threadpool 'http client'

scroix commented:
@justparking Thanks for the detailed dumps. I've pushed a refactor that replaces the per-node cached executor with a shared `ThreadPool("HTTP client", 16)`, so that thread growth should now be more linear.

It looks like this is the pattern for a couple of other subsystems (e.g. WebSocket, toolkit).

For validation, I ran Nodel with throwaway async test nodes and compared `jcmd $PID Thread.print` counts as well.

```python
# script.py

# Callback handlers
def on_complete(result):
    console.info('AsyncTest1: Request completed successfully')
    console.info('Response length: %s bytes' % len(result))

def on_error(exc):
    console.error('AsyncTest1: Request failed: %s' % exc)

# Trigger async HTTP request on startup
def main():
    console.warn('AsyncTest1: Triggering async HTTP request...')
    get_url('https://httpbingo.org/delay/1',
            async=True,
            complete=on_complete,
            error=on_error)
    console.warn('AsyncTest1: Request initiated (non-blocking)')
```

  ### Results

  | Async nodes | [Before](https://github.com/museumsvictoria/nodel/pull/362#issuecomment-3479289719) | After | Notes |
  |-------------|----------------------|-----------------|-------|
  | 0           | ~50 threads          | 63 threads      | measured baseline |
  | 1           | ~75 threads          | 72 threads      | measured |
  | 2           | ~102 threads         | 75 threads      | measured |
  | 10          | ~50 + 10×25 ≈ 300*    | 63 + 10×3 + 16 = 109[^calc] | projected |
  | 50          | ~50 + 50×25 ≈ 1300*   | 63 + 50×3 + 16 = 229 | projected |

[^calc]:  thread baseline, add 3 threads per async node. (two IO reactor + one “main”), then add 16 for the shared ThreadPool.

scroix committed:
Allow Content-Type headers with parameters so POSTs don’t fail before sending

scroix committed:
Merge branch 'dev' into feature/async_getURL

scroix committed:
refactor: rework async HTTP support addressing adversarial review

Correctness / reliability:
- bodyless responses (HEAD, 204, 304) no longer hang: response consumption
  now builds on HttpClient 5's AbstractAsyncResponseConsumer, which completes
  entity-less exchanges itself (content stays null, matching 4.x behaviour)
- close() no longer sleep-polls the JVM-wide connection counter (one node's
  restart could stall ~25s on other nodes' traffic); CloseMode.GRACEFUL
  already handles in-flight exchanges
- pre-emptive Basic Authorization is only sent for non-NTLM usernames again
  (DOMAIN\user credentials were leaking Basic-encoded), with an explicit
  charset; credential registration failures are logged instead of swallowed
- transparent gzip/deflate decompression restored (HttpClient 4 parity) and
  Accept-Encoding advertised unless the caller manages it
- async failures are never silent: without an error= callback the failure is
  raised into the node's console; callbacks receive the unwrapped cause
  rather than CompletionException; callback delivery failures are reported

Simplification:
- temp-file "streaming" deleted (it read the whole file back into memory at
  the end anyway); responses buffer in memory under the existing 150MB cap
- s_httpThreadPool deleted: the async client's execute() is non-blocking so
  it is called inline; active-connection counting is accurate again and
  decrements before completing futures
- NodelHTTPClient is async-first: makeRequestAsync is the abstract core and
  makeRequest/makeSimpleRequest(Async) derive from it with a single copy of
  the 2xx-check logic
- dead ManagedToolkit.getURLAsync and the reflection/JDK-only tests removed
- unused httpcore5-reactive dependency removed
- HTTPSimpleResponse.rawContent removed (receive rate is counted as bytes
  arrive); nodetoolkit.py whitespace churn reverted; get_url builds its
  argument list once

scroix committed:
Merge branch 'dev' into feature/async_getURL

scroix committed:
Merge upstream/dev into feature/async_getURL
